File: iOS/Objc/Examples/cases/AudioDiscernTask.m
#import <XCTest/XCTest.h>
#import <QCloudCOSXML/QCloudCOSXML.h>
#import <QCloudCOSXML/QCloudPostAudioDiscernTaskRequest.h>
#import <QCloudCOSXML/QCloudBatchGetAudioDiscernTaskRequest.h>
#import <QCloudCOSXML/QCloudGetAudioDiscernTaskRequest.h>
#import <QCloudCOSXML/QCloudPostAudioDiscernTaskInfo.h>
@interface AudioDiscernTask : XCTestCase <QCloudSignatureProvider, QCloudCredentailFenceQueueDelegate>

@property (nonatomic) QCloudCredentailFenceQueue* credentialFenceQueue;

@end

@implementation AudioDiscernTask

- (void)setUp {
    // 注册默认的 COS 服务
    QCloudServiceConfiguration* configuration = [QCloudServiceConfiguration new];
    configuration.appID = @"1253653367";
    configuration.signatureProvider = self;
    QCloudCOSXMLEndPoint* endpoint = [[QCloudCOSXMLEndPoint alloc] init];
    endpoint.regionName = @"ap-guangzhou";//服务地域名称，可用的地域请参考注释
    endpoint.useHTTPS = true;
    configuration.endpoint = endpoint;
    [QCloudCOSXMLService registerDefaultCOSXMLWithConfiguration:configuration];
    [QCloudCOSTransferMangerService registerDefaultCOSTransferMangerWithConfiguration:configuration];

    // 脚手架用于获取临时密钥
    self.credentialFenceQueue = [QCloudCredentailFenceQueue new];
    self.credentialFenceQueue.delegate = self;
}

- (void) fenceQueue:(QCloudCredentailFenceQueue * )queue requestCreatorWithContinue:(QCloudCredentailFenceQueueContinue)continueBlock
{
    QCloudCredential* credential = [QCloudCredential new];
    //在这里可以同步过程从服务器获取临时签名需要的 secretID，secretKey，expiretionDate 和 token 参数
    credential.secretID = @"COS_SECRETID";
    credential.secretKey = @"COS_SECRETKEY";
    credential.token = @"COS_TOKEN";
    /*强烈建议返回服务器时间作为签名的开始时间，用来避免由于用户手机本地时间偏差过大导致的签名不正确 */
    credential.startDate = [[[NSDateFormatter alloc] init] dateFromString:@"startTime"]; // 单位是秒
    credential.expirationDate = [[[NSDateFormatter alloc] init] dateFromString:@"expiredTime"];
    QCloudAuthentationV5Creator* creator = [[QCloudAuthentationV5Creator alloc]
        initWithCredential:credential];
    continueBlock(creator, nil);
}

- (void) signatureWithFields:(QCloudSignatureFields*)fileds
                     request:(QCloudBizHTTPRequest*)request
                  urlRequest:(NSMutableURLRequest*)urlRequst
                   compelete:(QCloudHTTPAuthentationContinueBlock)continueBlock
{
    [self.credentialFenceQueue performAction:^(QCloudAuthentationCreator *creator, NSError *error) {
        if (error) {
            continueBlock(nil, error);
        } else {
            QCloudSignature* signature =  [creator signatureForData:urlRequst];
            continueBlock(signature, nil);
        }
    }];
}

/**
 * 提交一个语音识别任务
 */
- (void)postAudioDiscernTask{
    //.cssg-snippet-body-start:[objc-post-audiodiscern]
    
    QCloudPostAudioDiscernTaskRequest * request = [[QCloudPostAudioDiscernTaskRequest alloc]init];

    // 存储桶名称，格式为 BucketName-APPID
    request.bucket = @"BucketName-APPID";
    request.regionName = @"regionName";

    QCloudPostAudioDiscernTaskInfo* taskInfo = [QCloudPostAudioDiscernTaskInfo new];
    taskInfo.Tag = @"SpeechRecognition";
    // 队列 ID ,通过查询语音识别队列获取
    taskInfo.QueueId = @"QueueId";
    // 操作规则
    QCloudPostAudioDiscernTaskInfoInput * input = QCloudPostAudioDiscernTaskInfoInput.new;
    input.Object = @"test1";
    // 待操作的语音文件
    taskInfo.Input = input;
    QCloudPostAudioDiscernOperation * op = [QCloudPostAudioDiscernOperation new];
    QCloudPostAudioDiscernTaskInfoOutput * output = QCloudPostAudioDiscernTaskInfoOutput.new;
    output.Region = @"regionName";
    output.Bucket = @"BucketName-APPID";
    output.Object = @"test";
    // 结果输出地址
    op.Output = output;

    QCloudPostAudioDiscernRecognition * speechRecognition = [QCloudPostAudioDiscernRecognition new];
    speechRecognition.EngineModelType =@"16k_zh";
    speechRecognition.ChannelNum = 1;
    speechRecognition.ResTextFormat = 0;
    speechRecognition.ConvertNumMode = 0;
    // 当 Tag 为 SpeechRecognition 时有效，指定该任务的参数
    op.SpeechRecognition = speechRecognition;
    // 操作规则
    taskInfo.Operation = op;
    //  语音识别任务
    request.taskInfo = taskInfo;

    [request setFinishBlock:^(QCloudPostAudioDiscernTaskResult * _Nullable result, NSError * _Nullable error) {
        // outputObject 提交审核反馈信息 包含用于查询的job id，详细字段请查看api文档或者SDK源码
        // QCloudPostAudioRecognitionResult 类；
    }];
    [[QCloudCOSXMLService defaultCOSXML] PostAudioDiscernTask:request];
    //.cssg-snippet-body-end
}

/**
 * 查询指定的语音识别任务
 */
- (void)getAudioDiscernTask {
    //.cssg-snippet-body-start:[objc-get-audiodiscern-task]
    QCloudGetAudioDiscernTaskRequest * request = [[QCloudGetAudioDiscernTaskRequest alloc]init];

    // 存储桶名称，格式为 BucketName-APPID
    request.bucket = @"examplebucket-1250000000";

    // QCloudPostAudioDiscernTaskRequest接口返回的jobid
    request.jobId = @"jobid";

    request.regionName = @"regionName";

    request.finishBlock = ^(QCloudGetAudioDiscernTaskResult * outputObject, NSError *error) {
        // outputObject 详细字段请查看api文档或者SDK源码
        // QCloudGetAudioDiscernTaskResult 类；
    };
    [[QCloudCOSXMLService defaultCOSXML] GetAudioDiscernTask:request];
    //.cssg-snippet-body-end
}

/**
 * 批量拉取语音识别任务
 */
-(void)batchGetAudioDiscernTask{
    //.cssg-snippet-body-start:[objc-batch-audiodiscern-task]
    QCloudBatchGetAudioDiscernTaskRequest * request = [[QCloudBatchGetAudioDiscernTaskRequest alloc]init];

    // 存储桶名称，格式为 BucketName-APPID
    request.bucket = @"examplebucket-1250000000";

    // 拉取该队列 ID 下的任务。
    // 通过查询语音识别队列获取
    request.queueId = @"queueId";

    request.regionName = @"regionName";

    request.states = QCloudTaskStatesSuccess | QCloudTaskStatesCancel;

    // 其他更多参数请查阅sdk文档或源码注释

    request.finishBlock = ^(QCloudBatchGetAudioDiscernTaskResult * outputObject, NSError *error) {
        // outputObject 任务结果，详细字段请查看api文档或者SDK源码
        // QCloudBatchGetAudioDiscernTaskResult 类；
    };
    [[QCloudCOSXMLService defaultCOSXML] BatchGetAudioDiscernTask:request];
    //.cssg-snippet-body-end
}

// .cssg-methods-pragma
- (void)testAudioOperation {
    // 提交一个语音识别任务
    [self postAudioDiscernTask];

    // 查询指定的语音识别任务
    [self getAudioDiscernTask];
    
    // 批量拉取语音识别任务
    [self batchGetAudioDiscernTask];
  
    // .cssg-methods-pragma
}

@end

File: iOS/Swift/Examples/cases/AudioDiscernTask.swift
import XCTest
import QCloudCOSXML
class AudioDiscernTask: XCTestCase,QCloudSignatureProvider,QCloudCredentailFenceQueueDelegate{

    var credentialFenceQueue:QCloudCredentailFenceQueue?;

    override func setUp() {
        let config = QCloudServiceConfiguration.init();
        config.signatureProvider = self;
        config.appID = "1253653367";
        let endpoint = QCloudCOSXMLEndPoint.init();
        endpoint.regionName = "ap-guangzhou";//服务地域名称，可用的地域请参考注释
        endpoint.useHTTPS = true;
        config.endpoint = endpoint;
        QCloudCOSXMLService.registerDefaultCOSXML(with: config);
        QCloudCOSTransferMangerService.registerDefaultCOSTransferManger(with: config);

        // 脚手架用于获取临时密钥
        self.credentialFenceQueue = QCloudCredentailFenceQueue();
        self.credentialFenceQueue?.delegate = self;
    }

    func fenceQueue(_ queue: QCloudCredentailFenceQueue!, requestCreatorWithContinue continueBlock: QCloudCredentailFenceQueueContinue!) {
        let cre = QCloudCredential.init();
        //在这里可以同步过程从服务器获取临时签名需要的 secretID，secretKey，expiretionDate 和 token 参数
        cre.secretID = "COS_SECRETID";
        cre.secretKey = "COS_SECRETKEY";
        cre.token = "COS_TOKEN";
        /*强烈建议返回服务器时间作为签名的开始时间，用来避免由于用户手机本地时间偏差过大导致的签名不正确 */
        cre.startDate = DateFormatter().date(from: "startTime"); // 单位是秒
        cre.expirationDate = DateFormatter().date(from: "expiredTime");
        let auth = QCloudAuthentationV5Creator.init(credential: cre);
        continueBlock(auth,nil);
    }

    func signature(with fileds: QCloudSignatureFields!, request: QCloudBizHTTPRequest!, urlRequest urlRequst: NSMutableURLRequest!, compelete continueBlock: QCloudHTTPAuthentationContinueBlock!) {
        self.credentialFenceQueue?.performAction({ (creator, error) in
            if error != nil {
                continueBlock(nil,error!);
            }else{
                let signature = creator?.signature(forData: urlRequst);
                continueBlock(signature,nil);
            }
        })
    }

    // 提交一个语音识别任务
    func postAudioDiscernTask() {
        //.cssg-snippet-body-start:[swift-post-audiodiscern]
        let request = QCloudPostAudioDiscernTaskRequest.init();

       // 存储桶名称，格式为 BucketName-APPID
       request.bucket = "BucketName-APPID";
       request.regionName = "regionName";

       let taskInfo = QCloudPostAudioDiscernTaskInfo.init();
       taskInfo.tag = "SpeechRecognition";
       
       // 队列 ID ,通过查询语音识别队列获取
       taskInfo.queueId = "QueueId";
       // 操作规则
       let input = QCloudPostAudioDiscernTaskInfoInput.init();
       input.object = "test1";
       // 待操作的语音文件
       taskInfo.input = input;
       let op = QCloudPostAudioDiscernTaskInfoOperation.init();
       let output = QCloudPostAudioDiscernTaskInfoOutput.init();
       output.region = "regionName";
       output.bucket = "BucketName-APPID";
       output.object = "test";
       // 结果输出地址
       op.output = output;

       let speechRecognition = QCloudPostAudioDiscernTaskInfoSpeechRecognition.init();
       speechRecognition.engineModelType = "16k_zh";
       speechRecognition.channelNum = 1;
       speechRecognition.resTextFormat = 0;
       speechRecognition.convertNumMode = 0;
       // 当 Tag 为 SpeechRecognition 时有效，指定该任务的参数
       op.speechRecognition = speechRecognition;
       // 操作规则
       taskInfo.operation = op;
       //  语音识别任务
       request.taskInfo = taskInfo;

        request.setFinish { outputObject, error in
            // outputObject 提交审核反馈信息 包含用于查询的job id，详细字段请查看api文档或者SDK源码
            // QCloudPostAudioRecognitionResult 类；
        };
       QCloudCOSXMLService.defaultCOSXML().postAudioDiscernTask(request);
        
        //.cssg-snippet-body-end
    }

    // 查询指定的语音识别任务
    func getAudioDiscernTask() {
        //.cssg-snippet-body-start:[swift-get-audiodiscern-task]
        let request = QCloudGetAudioDiscernTaskRequest.init();

        // 存储桶名称，格式为 BucketName-APPID
        request.bucket = "examplebucket-1250000000";

        // QCloudPostAudioDiscernTaskRequest接口返回的jobid
        request.jobId = "jobid";

        request.regionName = "regionName";

        request.setFinish { outputObject, error in
            // outputObject 详细字段请查看api文档或者SDK源码
            // QCloudGetAudioDiscernTaskResult 类；
        };
        QCloudCOSXMLService.defaultCOSXML().getAudioDiscernTask(request);
        //.cssg-snippet-body-end
    }
    
    // 批量拉取语音识别任务
    func batchGetAudioDiscernTask() {
        //.cssg-snippet-body-start:[swift-batch-audiodiscern-task]
        let request = QCloudBatchGetAudioDiscernTaskRequest.init();

        // 存储桶名称，格式为 BucketName-APPID
        request.bucket = "examplebucket-1250000000";

        // 拉取该队列 ID 下的任务。
        // 通过查询语音识别队列获取
        request.queueId = "queueId";

        request.regionName = "regionName";

        request.states = QCloudTaskStatesEnum(rawValue: QCloudTaskStatesEnum.success.rawValue | QCloudTaskStatesEnum.cancel.rawValue)!;

        // 其他更多参数请查阅sdk文档或源码注释

        request.setFinish { outputObject, error in
            // outputObject 任务结果，详细字段请查看api文档或者SDK源码
            // QCloudBatchGetAudioDiscernTaskResult 类；
        };
        QCloudCOSXMLService.defaultCOSXML().batchGetAudioDiscernTask(request);
        //.cssg-snippet-body-end
    }
    
    // .cssg-methods-pragma
    func testAudioOperation() {
        // 提交一个语音识别任务
        self.postAudioDiscernTask();

        // 查询指定的语音识别任务
        self.getAudioDiscernTask();
        
        // 批量拉取语音识别任务
        self.batchGetAudioDiscernTask();
    
        // .cssg-methods-pragma
    }
}
